File: solutions/0001-0100/0012-integer-to-roman/solution.cpp
#include "solution.hpp"

#include <array>
#include <utility>

namespace leetcode::p0012 {

std::string solve(int num) {
    static constexpr std::array<std::pair<int, const char*>, 13> romanSymbols{{
        {1000, "M"}, {900, "CM"}, {500, "D"}, {400, "CD"}, {100, "C"}, {90, "XC"}, {50, "L"},
        {40, "XL"}, {10, "X"}, {9, "IX"}, {5, "V"}, {4, "IV"}, {1, "I"},
    }};

    std::string result;

    for (const auto& [value, symbol] : romanSymbols) {
        while (num >= value) {
            result += symbol;
            num -= value;
        }
    }

    return result;
}

}  // namespace leetcode::p0012

File: solutions/0001-0100/0012-integer-to-roman/solution.ml
type error = Out_of_range of int

let min_num = 1
let max_num = 3999

let roman_symbols =
  [
    (1000, "M");
    (900, "CM");
    (500, "D");
    (400, "CD");
    (100, "C");
    (90, "XC");
    (50, "L");
    (40, "XL");
    (10, "X");
    (9, "IX");
    (5, "V");
    (4, "IV");
    (1, "I");
  ]

let rec build number = function
  | [] -> ""
  | (value, symbol) :: rest ->
      if number >= value then
        symbol ^ build (number - value) ((value, symbol) :: rest)
      else build number rest

let solve num =
  if num < min_num || num > max_num then Error (Out_of_range num)
  else Ok (build num roman_symbols)
